test/src/utils/differentiator.test.cpp

```diff
@@ -0,0 +1,328 @@
+// Functionality
+// -------------------------------------------------------------------------------------------------
+
+/**
+ * Struct used for test fixtures testing functionality.
+ * Constructor sets the values of the DataPoint objects.
+ * Helper functions defined to assist in testing
+ */
+struct DifferentiatorFunctionality : public ::testing::Test {
+ protected:
+  Differentiator<float> diff_test;
+  DataPoint<float> test_point;
+  DataPoint<float> second_point;
+  DataPoint<float> third_point;
+
+  void SetUp()
+  {
+    // Sets values of Data points
+    test_point.value = 10.5;
+    test_point.timestamp = 1;
+
+    second_point.value = 18.6;
+    second_point.timestamp = 2;
+
+    third_point.value= second_point.value;
+    third_point.timestamp = 3;
+  }
+
+  /**
+   * @brief Returns the difference between two data point values.
+  */
+  float deltaValue(DataPoint<float> point_1, DataPoint<float> prev_point)
+  {
+    return point_1.value - prev_point.value;
+  }
+
+  /**
+   * @brief Returns the difference between two data point timestamps
+  */
+  float deltaTime(DataPoint<float> point_1, DataPoint<float> prev_point)
+  {
+    return point_1.timestamp- prev_point.timestamp;
+  }
+
+  /**
+   * @brief Returns the gradient between two data points
+  */
+  float gradientOfPoints(DataPoint<float> point_1, DataPoint<float> prev_point)
+  {
+    return (deltaValue(point_1, prev_point))/((deltaTime(point_1, prev_point))/1e6);
+  }
+  void TearDown() {}
+};
+
+/**
+ * Test fixture used for determining whether initialisation works correctly.
+ * The value should always return 0
+*/
+TEST_F(DifferentiatorFunctionality, differentiatorInitialisedValue)
+{
+  ASSERT_EQ(diff_test.update(test_point).value, 0)
+    << "Initalised value should be 0";
+}
+
+/**
+ * Test fixture used for determining whether initialisation works correctly.
+ * The timestamp should always be returned unchanged
+*/
+TEST_F(DifferentiatorFunctionality, differentiatorInitialisedTimestamp)
+{
+  ASSERT_EQ(diff_test.update(test_point).timestamp, test_point.timestamp)
+    << "Timestamp of Data Point shouldnt change on update";
+}
+
+/**
+ * Test fixture used for checking the correctness of the gradient between two points.
+ * The Value of update should return the same as the gradient between the two points.
+ */
+TEST_F(DifferentiatorFunctionality, differentiatorGradient)
+{
+  diff_test.update(test_point);
+  ASSERT_FLOAT_EQ(diff_test.update(second_point).value, gradientOfPoints(second_point, test_point))
+    << "Derivative differs from the gradient between the two points";
+}
+
+/**
+ * Test fixture used for ensuring timestamp remains unchanged.
+ * Timestamp should always return unchanged whenever update is called.
+ */
+TEST_F(DifferentiatorFunctionality, differentiatorTimestampReturn)
+{
+  diff_test.update(test_point);
+  ASSERT_EQ(diff_test.update(second_point).timestamp, second_point.timestamp)
+    << "Timestamp should remain unchanged on Update";
```

Comment:
We should find a way of doing this better, maybe first encapsulate the error message in a variable, also maybe we can create a helper function for tests so that we can do the assertion and also pass the message as a parameter. Maybe changing the colour of the print would also be a good idea (red maybe)